PurchasesCoreSwift/Support/ManageSubscriptionModalHelper.swift

```diff
@@ -0,0 +1,166 @@
+                self.showAppleManageSubscriptions(managementURL: managementURL, completion: completion)
+                return
+            }
+
+            self.openURL(managementURL, completion: completion)
+        }
+    }
+
+}
+
+@available(iOS 9.0, *)
+@available(macOS 10.12, *)
+@available(watchOS, unavailable)
+@available(tvOS, unavailable)
+private extension ManageSubscriptionsModalHelper {
+
+    func showAppleManageSubscriptions(managementURL: URL,
+                                      completion: @escaping (Result<Void, ManageSubscriptionsModalError>) -> Void) {
+#if os(iOS)
+        if #available(iOS 15.0, *) {
+            Task.init {
+                let result = await self.showSK2ManageSubscriptions()
+                completion(result)
+            }
+            return
+        }
+#endif
+        self.openURL(managementURL, completion: completion)
+    }
+
+    func openURL(_ url: URL, completion: @escaping (Result<Void, ManageSubscriptionsModalError>) -> Void) {
+#if os(iOS)
+        self.openURLIfNotAppExtension(url: url)
+#elseif os(macOS)
+        NSWorkspace.shared.open(url)
+#endif
+        completion(.success(()))
+    }
+
+#if os(iOS)
+    // we can't directly reference UIApplication.shared in case this SDK is embedded into an app extension.
+    // so we ensure that it's not running in an app extension and use selectors to call UIApplication methods.
+    func openURLIfNotAppExtension(url: URL) {
+        guard !systemInfo.isAppExtension,
+              let application = systemInfo.sharedUIApplication else { return }
+
+        if #available(iOS 10.0, *) {
+            // NSInvocation is needed because the method takes three arguments
+            // and performSelector works for up to 2
+            typealias ClosureType = @convention(c) (AnyObject, Selector, NSURL, NSDictionary?, Any?) -> Void
+
+            let selector: Selector = NSSelectorFromString("openURL:options:completionHandler:")
+            let methodIMP: IMP! = application.method(for: selector)
+            let openURLMethod = unsafeBitCast(methodIMP, to: ClosureType.self)
+            openURLMethod(application, selector, url as NSURL, nil, nil)
+        } else {
+            let selector = NSSelectorFromString("openURL:")
+            systemInfo.sharedUIApplication?.perform(selector, with: url)
+        }
+    }
+
+    @MainActor
+    @available(iOS 15.0, *)
+    @available(macOS, unavailable)
+    func showSK2ManageSubscriptions() async -> Result<Void, ManageSubscriptionsModalError> {
+        guard let application = systemInfo.sharedUIApplication,
+              let windowScene = application.currentWindowScene else {
+                  return .failure(.couldntGetWindowScene)
+        }
+
+        do {
+#if os(iOS) || os(macOS)
+            try await AppStore.showManageSubscriptions(in: windowScene)
+            return .success(())
+#else
+            fatalError("tried to call AppStore.showManageSubscriptions in a platform that doesn't support it!")
```

Comment:
this isn't strictly necessary because of the availability checks, but I think if you're on tvOS it'll still try to compile the unavailable methods and it'll produce an error